src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.StringSegment.cs

```diff
@@ -0,0 +1,496 @@
+            {
+                WriteStringSegmentIndentedProlog();
+            }
+            else
+            {
+                WriteStringSegmentMinimizedProlog();
+            }
+        }
+
+        private void WriteStringSegmentIndentedProlog()
+        {
+            int indent = Indentation;
+            Debug.Assert(indent <= _indentLength * _options.MaxDepth);
+
+            // One quote and optionally 1 indent, 1 list separator and 1-2 bytes for new line
+            int bytesRequired = 1 + indent + 1 + _newLineLength;
+            if (_memory.Length - BytesPending < bytesRequired)
+            {
+                Grow(bytesRequired);
+            }
+
+            Span<byte> output = _memory.Span;
+
+            if (_currentDepth < 0)
+            {
+                output[BytesPending++] = JsonConstants.ListSeparator;
+            }
+
+            if (_tokenType != JsonTokenType.PropertyName)
+            {
+                if (_tokenType != JsonTokenType.None)
+                {
+                    WriteNewLine(output);
+                }
+                WriteIndentation(output.Slice(BytesPending), indent);
+                BytesPending += indent;
+            }
+
+            output[BytesPending++] = JsonConstants.Quote;
+        }
+
+        private void WriteStringSegmentMinimizedProlog()
+        {
+            // One quote and optionally 1 list separator
+            int bytesRequired = 2;
+            if (_memory.Length - BytesPending < bytesRequired)
+            {
+                Grow(bytesRequired);
+            }
+
+            Span<byte> output = _memory.Span;
+
+            if (_currentDepth < 0)
+            {
+                output[BytesPending++] = JsonConstants.ListSeparator;
+            }
+
+            output[BytesPending++] = JsonConstants.Quote;
+        }
+
+        private void WriteStringSegmentEpilog()
+        {
+            if (_memory.Length == BytesPending)
+            {
+                Grow(1);
+            }
+
+            _memory.Span[BytesPending++] = JsonConstants.Quote;
+        }
+
+#if NET
+        [InlineArray(4)]
+        private struct Inline4ByteArray
+        {
+            public byte byte0;
+        }
+#endif
+
+        private void GetPartialUtf8CodePoint(out ReadOnlySpan<byte> codePointBytes)
+        {
+            ReadOnlySpan<byte> partialCodePointBytes = PartialCodePointRaw;
+            Debug.Assert(partialCodePointBytes.Length == 4);
+
+            byte length = partialCodePointBytes[3];
+            Debug.Assert(0 <= length && length <= 4);
+
+            codePointBytes = partialCodePointBytes.Slice(0, length);
+        }
+
+        private void GetPartialUtf16CodePoint(out ReadOnlySpan<char> codePointChars)
+        {
+            ReadOnlySpan<byte> partialCodePointBytes = PartialCodePointRaw;
+            Debug.Assert(partialCodePointBytes.Length == 4);
+
+            byte length = partialCodePointBytes[3];
+            Debug.Assert(length == 2 || length == 0);
```

Comment:
Other than this assert, do we have code that checks if the buffer does actually contain UTF-16 data?

Reply:
Currently no, but this will likely need to be added depending on what we decide for the first design decision on https://github.com/dotnet/runtime/pull/101356#issuecomment-2546345225. If we choose to throw or write a replacement character then I can repurpose the length byte to also store the discriminator for UTF-8/UTF-16/Base64.

Reply:
If it helps simplify the code, we can start by forbidding mixing matching of encodings. Assuming there is demand in the future the restriction can be relaxed.

Reply:
It didn't complicate the code too much to allow mixing and emit a replacement character(s) when changing to a new encoding if there was an existing partial code byte. With this approach we can use the existing behavior when there is no partial code byte.